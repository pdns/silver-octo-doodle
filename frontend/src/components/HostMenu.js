import React from 'react';
import PropTypes from 'prop-types';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import IconButton from '@material-ui/core/IconButton';
import MoreVert from '@material-ui/icons/MoreVert';
import AddIcon from '@material-ui/icons/Add';
import MagnifyIcon from 'mdi-material-ui/Magnify';

class HostMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = { anchorEl: null };
  }

  handleClick = (event) => {
    this.setState({ anchorEl: event.currentTarget });
  }

  handleClose = () => {
    this.setState({ anchorEl: null });
  }

  createHandler = () => {
    this.props.onCreate();
    this.handleClose();
  }

  viewHandler = () => {
    this.props.onView();
    this.handleClose();
  }

  render() {
    const { anchorEl } = this.state;

    return (
      <React.Fragment>
        <IconButton
          onClick={this.handleClick}
        >
          <MoreVert />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          open={Boolean(anchorEl)}
          onClose={this.handleClose}
        >
          <MenuItem onClick={this.createHandler}>
            <ListItemIcon>
              <AddIcon />
            </ListItemIcon>
            <ListItemText primary="Create target" />
          </MenuItem>
          <MenuItem onClick={this.viewHandler}>
            <ListItemIcon>
              <MagnifyIcon />
            </ListItemIcon>
            <ListItemText primary="View details" />
          </MenuItem>
        </Menu>
      </React.Fragment>
    );
  }
}

HostMenu.propTypes = {
  onCreate: PropTypes.func.isRequired,
  onView: PropTypes.func.isRequired,
};

export default HostMenu;
