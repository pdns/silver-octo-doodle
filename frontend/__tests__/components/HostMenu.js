/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HostMenu from '../../src/components/HostMenu';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('@material-ui/core/Menu', () => 'Menu');
jest.mock('@material-ui/core/MenuItem', () => 'MenuItem');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/ListItemIcon', () => 'ListItemIcon');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/icons/MoreVert', () => 'MoreVertIcon');
jest.mock('@material-ui/icons/Add', () => 'AddIcon');
jest.mock('mdi-material-ui/Magnify', () => 'MagnifyIcon');

describe('HostMenu', () => {
  let onCreateFunc;
  let onViewFunc;
  let wrapper;

  beforeEach(() => {
    onCreateFunc = jest.fn();
    onViewFunc = jest.fn();
    wrapper = shallow(<HostMenu onCreate={onCreateFunc} onView={onViewFunc} />);
  });

  it('should render correctly with required props', () => {
    const tree = renderer
      .create(<HostMenu onCreate={onCreateFunc} onView={onViewFunc} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should set anchorEl on Menu by clicking IconButton', () => {
    expect(wrapper.find('Menu').prop('anchorEl')).toBe(null);
    wrapper.find('IconButton').simulate('click', { currentTarget: 'yay' });
    expect(wrapper.find('Menu').prop('anchorEl')).toBe('yay');
  });

  it('should nullify anchorEl on Menu by calling function passed to Menu.onClose', () => {
    wrapper.find('IconButton').simulate('click', { currentTarget: 'yay' });
    wrapper.find('Menu').simulate('close');
    expect(wrapper.find('Menu').prop('anchorEl')).toBe(null);
  });

  it('should call onCreate prop and nullify anchorEl by clicking on MenuItem for add', () => {
    wrapper.find('IconButton').simulate('click', { currentTarget: 'yay' });
    wrapper.find('MenuItem').at(0).simulate('click');
    expect(onCreateFunc).toHaveBeenCalledTimes(1);
    expect(wrapper.find('Menu').prop('anchorEl')).toBe(null);
  });

  it('should call onView prop and nullify anchorEl by clicking on MenuItem for add', () => {
    wrapper.find('IconButton').simulate('click', { currentTarget: 'yay' });
    wrapper.find('MenuItem').at(1).simulate('click');
    expect(onViewFunc).toHaveBeenCalledTimes(1);
    expect(wrapper.find('Menu').prop('anchorEl')).toBe(null);
  });
});
