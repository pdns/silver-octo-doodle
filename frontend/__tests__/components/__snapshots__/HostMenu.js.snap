// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostMenu should render correctly with required props 1`] = `
Array [
  <IconButton
    onClick={[Function]}
  >
    <MoreVertIcon />
  </IconButton>,
  <Menu
    anchorEl={null}
    onClose={[Function]}
    open={false}
  >
    <MenuItem
      onClick={[Function]}
    >
      <ListItemIcon>
        <AddIcon />
      </ListItemIcon>
      <ListItemText
        primary="Create target"
      />
    </MenuItem>
    <MenuItem
      onClick={[Function]}
    >
      <ListItemIcon>
        <MagnifyIcon />
      </ListItemIcon>
      <ListItemText
        primary="View details"
      />
    </MenuItem>
  </Menu>,
]
`;
